packages/core/http/core-http-versioned-router-server-internal/src/core_versioned_route.ts

```diff
@@ -0,0 +1,168 @@
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type {
+  RequestHandler,
+  IRouter,
+  RequestHandlerContextBase,
+  KibanaRequest,
+  KibanaResponseFactory,
+} from '@kbn/core-http-server';
+import type {
+  Version,
+  AddVersionOpts,
+  VersionedRoute,
+  VersionedRouteConfig,
+} from '@kbn/core-http-server';
+import type { CoreKibanaRequest } from '@kbn/core-http-router-server-internal';
+import { Mutable } from 'utility-types';
+import type { Method } from './types';
+
+import { validate } from './validate';
+
+type Options = AddVersionOpts<unknown, unknown, unknown, unknown>;
+
+export const VERSION_HEADER = 'TBD';
+
+// This validation is a pass-through so that we can apply our version-specific validation later
+const passThroughValidation = { body: schema.any(), params: schema.any(), query: schema.any() };
+
+export class CoreVersionedRoute implements VersionedRoute {
+  private readonly handlers = new Map<
+    Version,
+    {
+      fn: RequestHandler;
+      options: Options;
+    }
+  >();
+
+  public static from({
+    router,
+    method,
+    path,
+    options,
+  }: {
+    router: IRouter;
+    method: Method;
+    path: string;
+    options: VersionedRouteConfig<Method>;
+  }) {
+    return new CoreVersionedRoute(router, method, path, options);
+  }
+
+  private constructor(
+    private readonly router: IRouter,
+    public readonly method: Method,
+    public readonly path: string,
+    public readonly options: VersionedRouteConfig<Method>,
+    // TODO: Make "true" dev-only
+    private readonly validateResponses: boolean = true
+  ) {
+    this.router[this.method](
+      {
+        path: this.path,
+        validate: passThroughValidation,
+        options: this.options,
+      },
+      this.requestHandler
+    );
+  }
+
+  private getAvailableVersionsMessage(): string {
+    return `Available versions are: "${[...this.handlers.keys()].join(',') || 'none'}"`;
+  }
+
+  /** This is where we must implement the versioned spec once it is available */
+  private requestHandler = async (
+    ctx: RequestHandlerContextBase,
+    req: KibanaRequest,
+    res: KibanaResponseFactory
+  ) => {
+    const version = req.headers[VERSION_HEADER] as undefined | Version;
+    if (!version) {
+      return res.custom({
+        statusCode: 406,
+        body: `Version expected at [${this.method}] [${
+          this.path
+        }]. Please specify a version using the "${VERSION_HEADER}" header. ${this.getAvailableVersionsMessage()}`,
+      });
+    }
+
+    const handler = this.handlers.get(version);
+    if (!handler) {
+      return res.custom({
+        statusCode: 406,
+        body: `No version "${version}" available for [${this.method}] [${
+          this.path
+        }]. ${this.getAvailableVersionsMessage()}`,
+      });
+    }
+
+    const validation = handler.options.validate || undefined;
+
+    const mutableCoreKibanaRequest = req as Mutable<CoreKibanaRequest>;
```

Comment:
(moving comment from previous location)

This is so that we can hide/remove unvalidated inputs after CoreKibanaRequest was constructed. Not super happy about this, but it seemed like the least trouble without doing one of:

(1) Exposing more knowledge of Router internals so that we can dynamically set validation schema (right now we just use IRouter)
(2) Changing our approach, integrating the versioned router functionality into the Router implementation more directly